### ISSUE TITLE
Using the LD LIBRARY PATH to load local libraries does not work on Linux

### ISSUE THREAD

**Issue by wli3** [dotnet/runtime#34711]
_From @sunliusi on Sunday, March 22, 2020 4:21:18 AM_

I need to use the local library for cross-platform. Local libraries have different implementations and Put them in different folders. 

it work fine in windows just like this:
Environment.SetEnvironmentVariable("PATH", path + ";" + Environment.GetEnvironmentVariable("PATH"));

it's not working in linux just like this:
Environment.SetEnvironmentVariable("LD_LIBRARY_PATH", dir + ":" + Environment.GetEnvironmentVariable("LD_LIBRARY_PATH"));

_Copied from original issue: dotnet/sdk#10957_

**Comment by wli3** [comment]
_From @sunliusi on Sunday, March 22, 2020 4:28:55 AM_

NativeLibrary doesn't work either. There are two local libraries, one dependent on the other. This works only if the local library has no dependencies.

 ```
var path1 = Path.Combine(assemblyDirectory, "unix", "testaa.so");
                var p1 = NativeLibrary.Load(path1);
                if (p1 == IntPtr.Zero)
                {
                    throw new ApplicationException("load assembly fail:" + path1);
                }


                var path = Path.Combine(assemblyDirectory, "unix", "test.so");
                var p = NativeLibrary.Load(path);
                if (p == IntPtr.Zero)
                {
                    throw new ApplicationException("load assembly fail:" + path);
                }
```

 ---> System.DllNotFoundException: Unable to load shared library 'test.so' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: ./testaa.so: cannot open shared object file: No such file or directory
   at System.Runtime.InteropServices.NativeLibrary.LoadFromPath(String libraryName, Boolean throwOnError)
   at System.Runtime.InteropServices.NativeLibrary.Load(String libraryPath)

**Comment by wli3** [comment]
_From @SourceSkyBoxer on Tuesday, March 24, 2020 9:48:41 AM_

Hello, 
I will tell you example with NativeLibrary()
 ```
       /*
         *  Valid library name for Linux of X11, Extensions, OpenGL and Vulkan
         */
        private const string libX11 = "libX11.so.6";
        private const string libXrender = "libXrender.so.0";
        private static IntPtr handle;

        static XLib()
        {
            // Important instance of TryGetExport()
            handle = NativeLibrary.Load(libX11, typeof(XLib).Assembly, 0);
        }

        /*
         *  Display *XOpenDisplay(char *displayname);
         */
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        private delegate IntPtr _XOpenDisplay(string displayname);
        public static Display XOpenDisplay(string displayname)
        {
            _XOpenDisplay opendisplay_delegate;
            if (NativeLibrary.TryGetExport(handle, "XOpenDisplay", out IntPtr opendisplay))
            {
                opendisplay_delegate = Marshal.GetDelegateForFunctionPointer<_XOpenDisplay>(opendisplay);
            }
            else
            {
                opendisplay_delegate = (string name) => { throw new NotImplementedException("Error: XOpenDisplay not found."); };
            }

            return new Display(opendisplay_delegate(displayname));
        }
```
If you want FullPath like this.
You should use "AppDomain.CurrentDomain.BaseDirectory"
like this:
```
            ...
            string rootdir = AppDomain.CurrentDomain.BaseDirectory;
            handle = NativeLibrary.Load(Path.Combine(rootdir, "unix", "sofile.so"), typeof(YourClass).Assembly, 0);
            ...
```

You should create delegate with UnmanagedFunctionPointer
```
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        private delegate void _Method(string sayhello);
```
Than you can try as public method if you try out.
```
        public static void Method(string sayhello)
        {
            _Method method_delegate;
            if (NativeLibrary.TryGetExport(handle, "Method", out IntPtr method_ptr))
            {
                method_delegate = Marshal.GetDelegateForFunctionPointer<_Method>(method_ptr);
            }
            else
            {
                method_delegate = (string name) => { throw new NotImplementedException("Error: Method not found."); };
            }

            method_delegate(sayhello);
        }
```

Just it is small example of my made.
I hope my solution helps you as well.

// EDIT: You mean multiple so files?
No problem for me 

```
            ...
            // multiple so files
            string[] sofiles =
            {
                "file1.so", "file2.so"
            };
            string rootdir = AppDomain.CurrentDomain.BaseDirectory;
            foreach (string sofile in sofiles)
            {
                handle = NativeLibrary.Load(Path.Combine(rootdir, "unix", sofile));
            }
            ...
```
It is example:
For old version of NetFX 4.5 picture.
![image](https://user-images.githubusercontent.com/57066679/77443749-91dad000-6deb-11ea-87e2-f72f3de85c5e.png)
It is very easy to understand like I made it.

It is simple for you. I hope that.

**Comment by wli3** [comment]
_From @sunliusi on Wednesday, March 25, 2020 4:29:01 AM_

Thanks @SourceSkyBoxer  for such a detailed explanation.

Follow the example above:
`// multiple so files
            string[] sofiles =
            {
                "file1.so", "file2.so"
            };
            string rootdir = AppDomain.CurrentDomain.BaseDirectory;
            foreach (string sofile in sofiles)
            {
                handle = NativeLibrary.Load(Path.Combine(rootdir, "unix", sofile));
            }` 

The problem is that NativeLibrary.Load will fail when file2.so references file1.so.

If you put file2.so and file1.so in the root directory, there is no problem

**Comment by wli3** [comment]
_From @sunliusi on Wednesday, March 25, 2020 4:30:22 AM_

System.DllNotFoundException: Unable to load shared library '.../unix/file2.so' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: ./file1.so: cannot open shared object file: No such file or directory

**Comment by jkoritzinsky** [comment]
I believe that `LD_LIBRARY_PATH` has to be set before the program starts (either exported into the environment or when it is started). We don't propagate environment variable changes back out to the system when they're changed in .NET.

**Comment by jkoritzinsky** [comment]
Related to/possible duplicate of https://github.com/dotnet/runtime/issues/9529

**Comment by janvorli** [comment]
> I believe that `LD_LIBRARY_PATH` has to be set before the program starts (either exported into the environment or when it is started). We don't propagate environment variable changes back out to the system when they're changed in .NET on Unix.

That's correct. And it is not a .NET problem and it is unrelated to propagation env variables settings into the system, setting LD_LIBRARY_PATH has effect only before a process launch for any application written in any language. That's how the dynamic linker works.

**Comment by ssa3512** [comment]
I am running into a similar issue with interop on linux. I have a nuget package that is distributing some native libraries that I am calling with P/Invoke. The libraries are in the bin folder as such:

```
bin/Debug/netcoreapp3.1/runtimes
├── linux-x64
│   └── native
│       ├── libfile1.so
│       ├── libfile2.so
│       └── libfile3.so
├── linux-x86
│   └── native
│       ├── libfile1.so
│       ├── libfile2.so
│       └── libfile3.so
├── win-x64
│   └── native
│       ├── file1.dll
│       ├── file2.dll
│       └── file3.dll
└── win-x86
    └── native
        ├── file1.dll
        ├── file2.dll
        └── file3.dll
```

file1 has a dependency on file2 and file3. The code runs fine on Windows x86 and x64, but fails on linux with this error:
```
The active test run was aborted. Reason: Test host process crashed : Unhandled exception.
System.DllNotFoundException: Unable to load shared library 'file1' or one of its dependencies. In
order to help diagnose loading problems, consider setting the LD_DEBUG environment variable:
libfile1: cannot open shared object file: No such file or directory
```

The code is calling into this library as such:
``` cs
[DllImport("file1")
public static extern void NativeMethod();
```

If I copy the files `libfile1.so`, `libfile2.so` and `libfile3.so` to `/usr/lib` the code executes fine. Is this related to the above issue? I would expect .NET core to include the appropriate runtimes folder (in this case `runtimes/linux-x64/native` in the search path for interop.

**Comment by sunliusi** [comment]
Put so in the root directory on linux, and put dll in custom directory on windows, it also works.

Set LD_LIBRARY_PATH before the program start is not a good idea. It's best to control it using code, just like Windows

**Comment by ssa3512** [comment]
It is my understanding that the /runtimes/[platform]/native folder is supposed to be supported as a delivery mechanism for native libraries https://docs.microsoft.com/en-us/nuget/create-packages/supporting-multiple-target-frameworks#architecture-specific-folders

The challenge seems to come in when those libraries have other dependencies also being distributed this way.

**Comment by emmenlau** [comment]
I'm also affected by this problem, and I've spent the past day trying to figure out what is going wrong. Reading this issue I still do not see the best solution (or workaround).

Why does .NET not resolve libraries in architecture specific folders (i.e. in `runtimes/linux-x64/native/`)?

**Comment by sweemer** [comment]
I have encountered this issue as well and agree with @ssa3512 and @emmenlau that the .NET runtime should be responsible for somehow properly adding all nupkg native libs and their dependencies to the runtime search path.

In the meantime, I did find a workaround that I would like to share. You can use the compiler option `-Wl,-rpath,"\$ORIGIN"` when compiling (linking) the library that you specify in `DllImport`, which will add the executable directory to the search path. Since `dotnet publish` puts the executable and all the native libs in the `publish` directory together, the libraries are all available on the search path; i.e. the `publish` directory.

If you are using CMake then you can add `set(CMAKE_BUILD_RPATH "$\{ORIGIN\}")` to your `CMakeLists.txt` file to accomplish the same thing.

**Comment by gfody** [comment]
I've encountered this while trying to load libicuuc.dylib on macOS - this library sideloads libicudata.dylib and on Windows we can use SetDllDirectory from kernel32 at runtime but on macOS nothing seems to work. Is recompiling the native libs to embed an rpath really the only way?

**Comment by Rikj000** [comment]
After a long search, I've discovered a feasible work-around! 🥳

With [`patchelf --set-rpath`](https://github.com/NixOS/patchelf) it's possible to update the `RUNPATH` of libraries that depend on each-other,   
which solves `ldd` reporting `not found` on the dependencies,   
and also the ability to import the libraries with `NativeLibrary.Load()`

I've written a small `.sh` script to patch all libraries used:
```sh
#!/bin/bash

# Notes:
# - Depends on 'patchelf'
# - Needs to be called once after each library update

RUNTIMES=(
    "linux-arm64"
    "linux-x64"
);

LIBRARIES=(
    "file1.so"
    "file2.so"
);
  
for RUNTIME in "${RUNTIMES[@]}"; do
    echo "Patching RUNPATHs for '$RUNTIME' runtime:";
    for LIBRARY in "${LIBRARIES[@]}"; do
        echo "- Patching '$LIBRARY' library";
        patchelf --set-rpath "runtimes/$RUNTIME/native" "$RUNTIME/$LIBRARY";
    done
    echo "";
done

echo "Patched all library RUNPATHs!";
```

The script assumes that the libraries will end up under `runtimes/$RUNTIME/native` after building.

```bash
# Before patching
readelf -d file1.so | grep RUNPATH
 0x000000000000001d (RUNPATH)            Library runpath: [./]
# After patching
 0x000000000000001d (RUNPATH)            Library runpath: [runtimes/linux-arm64/native]
```

**Comment by teikitel** [comment]
>
> After a long search, I've discovered a feasible work-around! 🥳
> 

Thanks a lot for the workaround @Rikj000. I recently faced the very same issue and was able to patch my lib's rpath using https://pypi.org/project/patchelf/

I also had to do the same for a MacOs native lib where I use `otool` to check and `install_name_tool` to patch the rpath https://stackoverflow.com/questions/35220111/install-name-tool-difference-between-change-and-id

**Comment by haras-unicorn** [comment]
im sorry - i dont like when someone comments like this but i am very frustrated by this
i just cant believe that a project of this size doesnt support something as basic as LD_LIBRARY_PATH but calls itself cross platform
on nixos this is a major hurdle because it basically amounts to me patching up the sdk any time i want to use a system library

**Comment by haras-unicorn** [comment]
```nix
    addSdkLibraries =
      {
        libraries,
        rewrites ? [ ],
      }:
      sdk:
      sdk.overrideAttrs (
        final: prev: {
          # NOTE: because of the wrapper
          src = prev.src.overrideAttrs (
            final: prev: {
              postBuild = (prev.postBuild or "") + ''
                for verDir in $out/share/dotnet/shared/Microsoft.AspNetCore.App/*; do
                  ${builtins.concatStringsSep "\n  " (
                    builtins.map (library: ''
                      ln -sf ${library}/lib/*.so "$verDir"
                    '') libraries
                  )}

                  ${builtins.concatStringsSep "\n  " (
                    builtins.map (rewrite: ''
                      ln -sf "$verDir/${rewrite.from}" "$verDir/${rewrite.to}"
                    '') rewrites
                  )}
                done

                for verDir in $out/share/dotnet/shared/Microsoft.NETCore.App/*; do
                  ${builtins.concatStringsSep "\n  " (
                    builtins.map (library: ''
                      ln -sf ${library}/lib/*.so "$verDir"
                    '') libraries
                  )}

                  ${builtins.concatStringsSep "\n  " (
                    builtins.map (rewrite: ''
                      ln -sf "$verDir/${rewrite.from}" "$verDir/${rewrite.to}"
                    '') rewrites
                  )}
                done
              '';
            }
          );
        }
      );
```

made this nix function that does the job if anyone else on nix runs into this

**Comment by janvorli** [comment]
> i just cant believe that a project of this size doesnt support something as basic as LD_LIBRARY_PATH

@haras-unicorn thank you for sharing your script. The problem is not that we don't support the LD_LIBRARY_PATH. We support it the same way as any other app can on Linux. The OS loader reads it only before the application starts. So if you set it before starting your app, then it works as expected. The problem reported in this issue was that it is not possible to set LD_LIBRARY_PATH from inside of your app and expect the app to honor it. That would not work even in any app written in plain C.

**Comment by jkotas** [comment]
The question has been answered, and I do not see anything other actionable here. Closing.
